equational-reasoning.md

```diff
@@ -0,0 +1,29 @@
+Equational reasoning
+====================
+
+**Tags**: *substitution model*, *referential transparency*, *purity*, *side effect*
+
+Concerning programming, we can say that a function definition limits the scope where different expressions are evaluated in order to produce a result. This result is produced by the transformation of its input parameters (if they exist) along with any other internal variables and defines the return type of that function.
```

Comment:
What are these internal variables. Do they vary between function calls? If not, why are they variables?

Reply:
Maybe could be expressed in a nicer way, but what I want to say is that inside the scope of a function we can define some variables too. So at the end the result of the function is the result of transformation of these variables and the function parameters . I though that that was kind of interesting to mention so that one can understand what kind of transformation is provided by a function that takes no parameters

Reply:
what Dimitrios means are name assignments to inner subexpressions, like `val`s inside functions in Scala or when in Haskell we use `let .. in` or `where` inside the function

Reply:
here `variable` is an ambiguous term, because it might be a position in memory which you can reasign its value, or a label over a subexpression which is immutable